### ISSUE TITLE
HTMLPopupElement - <popup>

### ISSUE THREAD

**Issue by mfreed7** [w3ctag/design-reviews#680]
Braw mornin' TAG! (That's a new one for me 😄 )

I'm requesting a TAG review of the `<popup>` element. This is effectively a re-opening of the [previously requested TAG review for this feature](https://github.com/w3ctag/design-reviews/issues/599).

That review was closed, waiting for the proposed "Anchor Positioning" feature to be available for review. Please see [this comment](https://github.com/w3ctag/design-reviews/issues/599#issuecomment-863632279) on the old review, linking to the explainer for Anchor Positioning, plus some additional context. Essentially, the Anchor Positioning feature now has an explainer and a significant amount of brainstorming is happening around the [shape of the API](https://docs.google.com/presentation/d/1g0kCtpbGHqzJybhrP1vgbQapXSZW3zMopYOjNfBf0OQ/edit#slide=id.p). From those links you should be able to get a sense for what will be possible, positioning-wise, for `<popup>`. And hopefully that's enough to re-open and continue the TAG review for `<popup>`. In the time since the last TAG review, a [draft specification has been written](https://open-ui.org/components/popup) and reviewed by editors of the WHATWG HTML spec. And a prototype implementation in Chromium is basically complete. Several of the comments on the prior TAG review lead to changes that have been made in the spec and implementation, e.g. the addition of an `initiallyopen` attribute to declaratively open the popup.

  - Explainer¹: https://open-ui.org/components/popup.research.explainer
  - Specification URL: https://open-ui.org/components/popup
  - Tests: https://wpt.fyi/results/html/semantics/interactive-elements/the-popup-element
  - User research: https://open-ui.org/components/popup.research
  - Security and Privacy self-review²: 
  - GitHub repo (if you prefer feedback filed there): 
  - Primary contacts (and their relationship to the specification):
      - Mason Freed (@mfreed7), Google (current feature champion)
      - Melanie Richards (@melanierichards), Netlify, previously Microsoft (previous champion, wrote spec draft)
  - Organization(s)/project(s) driving the specification: OpenUI, Google, Microsoft
  - Key pieces of existing multi-stakeholder review or discussion of this specification: [Prior TAG review](https://github.com/w3ctag/design-reviews/issues/599)
  - External status/issue trackers for this specification: https://github.com/openui/open-ui/issues?q=is%3Aissue+is%3Aopen+label%3Apopup

Further details:

  - [X] I have reviewed the TAG's [Web Platform Design Principles](https://w3ctag.github.io/design-principles/)
  - Relevant time constraints or deadlines: None
  - The group where the work on this specification is currently being done: [OpenUI](https://open-ui.org/)
  - The group where standardization of this work is intended to be done: WHATWG HTML/DOM
  - Major unresolved issues with or opposition to this specification: None that I know of
  - This work is being funded by: Google & Microsoft

You should also know that...

I appreciate the TAG. 👍 

We'd prefer the TAG provide feedback as (please delete all but the desired option):

  💬 leave review feedback as a **comment in this issue** and @-notify @mfreed7

**Comment by mfreed7** [comment]
> > A `<popup>` is something that has these qualities:
> 
> That doesn't really help. Those are behaviors, not semantics or types of controls. (I.e. the things that feed into roles.)
> 
> Below you propose that the role should be dialog. That implies that instead of being a new element, this should be some sort of mode of the dialog element, since the dialog element already occupies the semantic space of dialogs. (E.g., `<dialog type="light-dismiss">` or `dialogEl.showWithLightDismiss()`.)
> 
> > In general, the `<dialog>` element is already in a somewhat similar boat, no? I.e. a `<dialog>` can be used for a number of UI purposes, such as dialogs, tooltips, toasts, etc. But the platform uses role=dialog and it's up to the developer to specialize further with ARIA attributes if it is being used for a more-specific purpose such as a tooltip.
> 
> No, this isn't correct. The dialog element can only be used for dialogs. Tooltips or toasts should be done using `<div>`s or similar. (Or perhaps a new element, if someone were bold enough to propose one... I seem to recall something of that sort for toasts.) If you use `<dialog>` but then override its exposure to accessibility tech using `role=something else`, the element is being misused.

So we both understand that `<dialog>` can be, and is, used for things other than dialogs. It is the (currently only) platform feature that gives developers access to the top layer. Using a `<div>` requires gymnastics like re-parenting the `<div>` to the last child of `<body>`, and giving it `z-index:9999999999`. It is also sub-par from an accessibility POV for many reasons, including the aforementioned re-parenting, plus the fact that `<div>`/`<span>` soup doesn't get announced at all by AT, right? At least if you build a toast using `<dialog>`, it'll be announced as a dialog, which isn't terribly far from a toast. Or at least it's closer than a `<div>`.

Having said that, I'm not disagreeing with the idea that we should make this semantically as clear and accessible as possible. I'm open to concrete proposals about how to do that.

> 
> We need to figure out what the similar semantic space, and corresponding role, is for popup. The explainer example seems to give several types of controls: menus, form element suggestions, content pickers, and teaching UI. What is the ARIA role for each of these? Perhaps each should have a separate element, not `<popup>`, which implements the appropriate semantics (and behaviors?? I can't imagine a teaching UI needs the exact same behaviors as a menu). Or perhaps some of those are better suited for `<dialog>` instead of `<popup>`.
> 
> I apologize that this is rather fundamental criticism, but I do feel like I've raised it repeatedly since the beginning of this endeavor, without much response. I appreciate getting some concrete engagement at this point.

No problem on the criticism - I'd like to get to a good solution here. I apologize that I don't think I realized this was "fundamental" - it sounds like you'd object to a general purpose `<popup>` element that fills many roles. So are you saying you'd prefer to see a proposal for many new elements, all with the same basic behavior (the three qualities I mentioned in my last comment), but different accessibility roles and semantic meaning? E.g. `<tooltip>`, `<listbox>` (would be useful in `<selectmenu>`), `<menu>` (oops), etc.?

I filed https://github.com/openui/open-ui/issues/410 to discuss this issue further. Also happy to continue the discussion here.

**Comment by domenic** [comment]
> So are you saying you'd prefer to see a proposal for many new elements, all with the same basic behavior (the three qualities I mentioned in my last comment), but different accessibility roles and semantic meaning?

Yeah, that would be one approach. Or `<popup type="x">`, but then you have a defaulting issue...

It's also worth considering how to map accessibility roles to use cases, e.g. the explainer's menus, form element suggestions, content pickers, and teaching UI. It's not clear to me whether a teaching UI is one of those, for example...

And again I'll reiterate that it's a bit surprising that those three exact behaviors (and no further behaviors) perfectly encapsulate what developers want for menus, listboxes, tooltips, teaching UIs, content pickers, and form element suggestions. E.g. at least on every operating system I'm familiar with menus and tooltips have very different focus behaviors, if nothing else.

**Comment by mfreed7** [comment]
> > So are you saying you'd prefer to see a proposal for many new elements, all with the same basic behavior (the three qualities I mentioned in my last comment), but different accessibility roles and semantic meaning?
> 
> Yeah, that would be one approach. Or `<popup type="x">`, but then you have a defaulting issue...
> 
> It's also worth considering how to map accessibility roles to use cases, e.g. the explainer's menus, form element suggestions, content pickers, and teaching UI. It's not clear to me whether a teaching UI is one of the three you listed (tooltip, listbox, menu), for example...
> 
> And again I'll reiterate that it's a bit surprising that those three exact behaviors (and no further behaviors) perfectly encapsulate what developers want for menus, listboxes, tooltips, teaching UIs, content pickers, and form element suggestions. E.g. at least on every operating system I'm familiar with menus and tooltips have very different focus behaviors, if nothing else.

These are very good points - there will definitely be some behavioral differences among these element types. I wasn't saying all of these elements would behave the same, I was saying that they (might?) all share the three "basic" behaviors with `<popup>`, in addition to some others. Or perhaps even some tweaks to the three.

Let's [continue the discussion within the issue](https://github.com/openui/open-ui/issues/410) - there will likely be more people who can weigh in.

**Comment by mfreed7** [comment]
@plinss thanks for the comments. Have you seen [the research page for `<popup>`](https://open-ui.org/components/popup.research) on OpenUI? If not, I'd encourage you to take a look. I do think the approach is being fairly well thought-out. Perhaps not satisfyingly, the answer is "both". As you'll see on that page, a `<popup>` can be defined fairly broadly, and in significant detail, as an element that can be used on its own. It also forms a "base class" for several more-specific types of UI that share the base behaviors of `<popup>` but add or modify them in various ways. Those sub-class elements have high level semantics that map closely to ARIA roles. The super-class is more generic, and is basically "a lightweight, ephemeral piece of UI content that is displayed on top of page content".

**Comment by LeaVerou** [comment]
There is precedent in the Web Platform for having an abstract base class with shared API and subclasses that correspond to specific tag names: `HTMLMediaElement` contains the shared behaviors between `<audio>` and `<video>`, whereas the base class does not correspond to a tag name, because it cannot be used directly. I wonder if a similar architecture would be more appropriate here than introducing a new element with no concrete semantics?

**Comment by hober** [comment]
> There is precedent in the Web Platform for having an abstract base class that contains shared API, and subclasses that correspond to specific tag names: `HTMLMediaElement` contains the shared behaviors between `<audio>` and `<video>`, whereas the base class does not correspond to a tag name, because it cannot be used directly. I wonder if a similar architecture would be more appropriate here than introducing a new element with no concrete semantics?

Indeed, and should that base class also be the base class of `<dialog>`?

**Comment by hober** [comment]
Hi everybody! Thank you for your patience as we've taken a while to get back to you on this one. Special thanks to @domenic for his thorough review of this feature. We're happy to see that there's been a bunch of productive conversation on this here and in various OpenUI issues. It sounds like the direction you expect this to go in is pretty different from what the current documents describe. We'd like to hold off on reviewing until you've updated the documents to match the current understanding. Please let us know when to revisit this!
